### PR TITLE
MULE-11573 Migrate 'synchronous' processing strategy away from using …

### PR DESCRIPTION
**Pull request by dfeist**
…blocking code path

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java [hunk 1]
```diff
@@ -6,39 +6,10 @@
...
-      return event -> {
-      };
-    }
-  };
-
-  @Override
-  public ProcessingStrategy create(MuleContext muleContext, String schedulersNamePrefix) {
-    return LEGACY_SYNCHRONOUS_PROCESSING_STRATEGY_INSTANCE;
-  }
+public class LegacySynchronousProcessingStrategyFactory extends CallerThreadProcessingStrategyFactory {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java]: do we need to keep this?

File: tests/unit/src/main/java/org/mule/tck/MuleTestUtils.java [hunk 2]
```diff
@@ -56,7 +57,9 @@ public static Injector spyInjector(MuleContext muleContext) {
   }
 
   public static Flow getTestFlow(MuleContext context) throws MuleException {
-    final Flow flow = builder(APPLE_FLOW, context).build();
+    // Create test flow with processing strategy that doesn't create stream infrastructure.
+    final Flow flow =
+        builder(APPLE_FLOW, context).processingStrategyFactory(new CallerThreadProcessingStrategyFactory()).build();
```
Comment (elrodro83) [inline tests/unit/src/main/java/org/mule/tck/MuleTestUtils.java]: don't we lose reactor coverage if we make this change in the default?

File: core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java (hunk 1 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java]: Will cleanup after M5, spec and discussion. For now i'm removing it's use when `processingStrategy="synchronous"` is configured.

File: core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java [hunk 3]
```diff
@@ -6,39 +6,20 @@
...
-      return event -> {
-      };
-    }
-  };
-
   @Override
   public ProcessingStrategy create(MuleContext muleContext, String schedulersNamePrefix) {
-    return LEGACY_SYNCHRONOUS_PROCESSING_STRATEGY_INSTANCE;
+    return (flowConstruct, pipelineFunction) -> {
+      throw new IllegalStateException("Sink cannot be created for " + LegacySynchronousProcessingStrategyFactory.class.getName());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java]: better UnsupportedOperationException?

File: core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      @Override
+      public void accept(Event event) {
+        onEventConsumer.accept(event);
+        just(event).transform(function).subscribe();
+      }
+
+    };
+  }
+
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java]: why so many linebreaks? :P

File: core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java (hunk 3 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/LegacySynchronousProcessingStrategyFactory.java]: In this case i think illegalState is more specific/correct, as this should never be called do it if conditions in AbstractPipeline/Flow

File: core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java (hunk 4 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java]: Fixes

File: core-tests/src/test/java/org/mule/runtime/core/processor/strategy/AbstractProcessingStrategyTestCase.java [hunk 5]
```diff
@@ -93,7 +94,7 @@ public void before() throws RegistrationException {
     cpuLight = new TestScheduler(3, CPU_LIGHT);
     blocking = new TestScheduler(3, IO);
     cpuIntensive = new TestScheduler(3, CPU_INTENSIVE);
-    asyncExecutor = newSingleThreadExecutor();
+    asyncExecutor = newCachedThreadPool();
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/processor/strategy/AbstractProcessingStrategyTestCase.java]: why not use a Scheduler for this?

File: core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import org.reactivestreams.Publisher;
+
+/**
+ * Abstract processing strategy that, rather than using a single shared stream, creates a stream per {@link Event}.
+ */
+abstract class AbstractStreamPerEventProcessingStrategyFactory extends AbstractProcessingStrategy {
+
+  @Override
+  public Sink createSink(FlowConstruct flowConstruct, Function<Publisher<Event>, Publisher<Event>> function) {
+    return new Sink() {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java]: other than this, don't we have some common sink implementations worth extracting to an abstract class?

File: core/src/main/java/org/mule/runtime/core/processor/strategy/SynchronousProcessingStrategyFactory.java [hunk 7]
```diff
@@ -14,26 +14,30 @@
 import java.util.function.Consumer;
 
 /**
- * This processing strategy processes all message processors in the calling thread.
+ * Processing strategy that processes all message processors in the caller thread. Unlike other, asynchronous, processing
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/SynchronousProcessingStrategyFactory.java]: replace "message processors" with a link to Processor

File: core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.reactivestreams.Publisher;
+
+/**
+ * Abstract processing strategy that, rather than using a single shared stream, creates a stream per {@link Event}.
+ */
+abstract class AbstractStreamPerEventProcessingStrategyFactory extends AbstractProcessingStrategy {
+
+  @Override
+  public Sink createSink(FlowConstruct flowConstruct, Function<Publisher<Event>, Publisher<Event>> function) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/strategy/AbstractStreamPerEventProcessingStrategyFactory.java]: make final, review other cases

**Pull request merged**